src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -47,8 +47,8 @@ Policy::LookupInfo Policy::lookup(const std::string &global_image_id) {
   return lookup(global_image_id, m_map_lock);
 }
 
-bool Policy::add_image(const std::string &global_image_id,
-                       Context *on_update, Context *on_acquire, Context *on_finish) {
+bool Policy::add_image(const std::string &global_image_id, Context *on_update,
+                       Context *on_acquire, Context *on_remove, Context *on_finish) {
```

Comment:
Why would ```ImageMap``` pass a handler to remove this item from on-disk? This transition is really an internal state cleanup, so when the on-disk update fails, the ```Policy``` in-memory map should be updated. There is no need to add the image to the "remove from on-disk map" queue.

Reply:
Wouldn't the on-disk map become out-of-sync with the in-memory map in this case? Although that wouldn't be incorrect functionally since we'll either update the map (on shuffle) or purge it later (on remove via the `UNASSIGNED -> REMOVE_MAPPING` state transition).

But, I think if we can, we should cleanup the on-disk state too. I agree that passing `on_remove` to `Policy::add_image` makes the interface a bit unclean.

Reply:
... but if the on-disk update failed, then there is no item to remove.

Reply:
on-disk updates are retried on failure, so we do have the on-disk map to remove.

Reply:
... so if it's just for cleaning up a failed assignment to a dead instance, it's probably not worth the effort to update on disk (for clarity) since it will just be immediately re-assigned to another instance. The "failure" to assign should just re-start the mapping process.

Reply:
yeh, makes sense.